feat: add cli flags for storage capacity per subnetwork

morph-dev opened the pull request:
### What was wrong?

Based on #1548 and discussion in the last meeting, we agreed that we following cli flags regarding storage capacity:

1. `storage.total` flag that specifies total storage capacity (in MB)
    - this capacity should be shared between all enabled networks
    - current flag `mb` should be set as alias (until we migrate in other places)
2. `storage.beacon`, `storage.history` and `storage.state` that specifies storage capacity (in MB) per subnetwork explicitly

Some restrictions:
- `storage.total` can't be combined with other flags
    - other flags can/should be used together, see next point
- if `storage.{subnetwork}` is specified for one enabled subnetwork, it should be specified for all enabled networks

For future work:
- if user didn't enable subnetwork, but we detect that it's using some storage on disk (e.g. from previous run), print warning message (or do something else, TBD).

### How was it fixed?

- created cli flags
    - added detailed documentation on how they should be used
- added checks when `TrinConfig` is created
    -  added tests
- adapted logic for creating `PortalStorageConfig`
    - added tests

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

KolbyML reviewed (approved):
@morph-dev over all the PR looks good

carver commented on trin-storage/src/config.rs:
```diff
@@ -76,3 +111,121 @@ pub struct PortalStorageConfig {
+    #[case::beaconhistorystate_beacon(vec![Subnetwork::Beacon, Subnetwork::History, Subnetwork::State], 100, Subnetwork::Beacon, Some(0))]
+    #[case::beaconhistorystate_history(vec![Subnetwork::Beacon, Subnetwork::History, Subnetwork::State], 100, Subnetwork::History, Some(1_000_000))]
+    #[case::beaconhistorystate_state(vec![Subnetwork::Beacon, Subnetwork::History, Subnetwork::State], 100, Subnetwork::State, Some(99_000_000))]
+    fn combined_capacity_config(
```
nice coverage 👍🏻

carver commented on ethportal-api/src/types/cli.rs:
```diff
@@ -312,8 +357,77 @@ impl TrinConfig {
+                            ),
+                        ));
+                    }
+                    if storage_flag.is_some() && !config.portal_subnetworks.contains(&subnetwork) {
+                        return Err(Error::raw(
+                            ErrorKind::ArgumentConflict,
+                            format!(
+                                "--storage.{} is set but {subnetwork} subnetwork is not enabled",
+                                subnetwork.to_cli_arg(),
+                            ),
+                        ));
+                    }
```
I'm not sure we *have* to error out in this case. I do like the impulse to be strict to avoid accidents.

But I can also imagine a scenario where I'm flipping between enabling and disabling a network, and having to do that in two places in the command being a little annoying.

I'm slightly in favor of dropping this if you're neutral.

carver commented on ethportal-api/src/types/cli.rs:
```diff
@@ -256,7 +301,7 @@ impl TrinConfig {
     pub fn new_from<I, T>(args: I) -> Result<Self, clap::Error>
     where
-        I: Iterator<Item = T>,
+        I: IntoIterator<Item = T>,
```
I do really like this change. It also added noise to a big PR. I think it belongs in its own commit/PR.

carver commented on ethportal-api/src/types/cli.rs:
```diff
@@ -709,6 +804,247 @@ mod test {
+            assert_eq!(
+                config.storage_capacity_config(),
+                StorageCapacityConfig::Specific {
+                    beacon_mb: Some(0),
```
I don't think there are any actual bugs here, but as a practice the test for setting a specific value should probably pick one that isn't the value you would arrive at by default.

carver reviewed (approved):
Love all the tests!

*nit* I just did a quick grep, and didn't see any tests that the `--mb` alias still works. I'd be in favor of a single simple one to make sure we don't accidentally disable it in the future.

morph-dev commented:
Rebased, fixed default capacity tests, added `--mb` and zero capacity tests, and reverted partially reverted refactoring changes. 
I will merge it like this.

Let me know if you have other comments and I will address them in the followup PR.
